tools/p4mlir-translate/translate.cpp

```diff
@@ -1115,6 +1143,63 @@ void P4HIRConverter::postorder(const P4::IR::ReturnStatement *ret) {
     }
 }
 
+mlir::Value P4HIRConverter::emitHeaderBuiltInMethod(const P4::BuiltInMethod *builtInMethod,
+                                                    mlir::OpBuilder &builder, mlir::Location loc) {
+    mlir::Value callResult;
+    auto header = resolveReference(builtInMethod->appliedTo);
+    auto validityBitRef =
+        builder.create<P4HIR::StructExtractRefOp>(loc, header, P4HIR::HeaderType::validityBit);
+    if (builtInMethod->name == P4::IR::Type_Header::setValid ||
+        builtInMethod->name == P4::IR::Type_Header::setInvalid) {
+        auto validVal =
+            getValidHeaderConstant(loc, builtInMethod->name == P4::IR::Type_Header::setInvalid
+                                            ? P4HIR::ValidityBit::Invalid
+                                            : P4HIR::ValidityBit::Valid);
+        builder.create<P4HIR::AssignOp>(loc, validVal, validityBitRef);
+    } else if (builtInMethod->name == P4::IR::Type_Header::isValid) {
+        auto validityBit = builder.create<P4HIR::ReadOp>(loc, validityBitRef);
+        callResult = builder.create<P4HIR::CmpOp>(loc, P4HIR::CmpOpKind::Eq, validityBit,
+                                                  getValidHeaderConstant(loc));
+    } else {
+        BUG("Unsupported builtin method: %1%", builtInMethod->name);
+    }
+
+    return callResult;
+}
+
+mlir::Value P4HIRConverter::emitHeaderUnionBuiltInMethod(
+    const P4::IR::Type_HeaderUnion *p4cHeaderUnionType, const P4::BuiltInMethod *builtInMethod,
+    mlir::OpBuilder &builder, mlir::Location loc) {
+    if (builtInMethod->name != P4::IR::Type_Header::isValid)
+        BUG("Unsupported Header Union builtin method: %1%", builtInMethod->name);
+
+    auto isHeaderUnionValid = getBoolConstant(loc, false);  // Start with a false value
+    auto headerUnion = resolveReference(builtInMethod->appliedTo);
+    auto dialectHeaderUnionType =
+        mlir::cast<P4HIR::HeaderUnionType>(getOrCreateType(p4cHeaderUnionType));
+    auto huFields = dialectHeaderUnionType.getFields();
+
+    for (auto huField : huFields) {
+        auto headerRef = builder.create<P4HIR::StructExtractRefOp>(loc, headerUnion, huField.name);
+        auto validityBitRef = builder.create<P4HIR::StructExtractRefOp>(
```

Comment:
Maybe you can refactor the code a bit, so you could reuse some pieces of `emitHeaderBuiltInMethod`?

E.g.:
 - Instead of passing `builtInMethod`, pass only method name
 - Instead of emitting a reference to a header (via `resolveReference`), inside do it outside, and pass the `mlir::Value` as argument.

This way you'd essentially need to call such refactored `emitHeaderBuiltInMethod` with `headerRef` argument and then combine the resulting values properly.